File: remoter/src/main/java/remoter/compiler/builder/BinderParamBuilder.java
package remoter.compiler.builder;

import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.MethodSpec;

import javax.annotation.processing.Messager;
import javax.lang.model.element.Element;
import javax.lang.model.element.VariableElement;
import javax.lang.model.type.ArrayType;
import javax.lang.model.type.TypeKind;
import javax.lang.model.type.TypeMirror;

import static com.google.auto.common.MoreElements.getPackage;

/**
 * A {@link ParamBuilder} for Binder type parameters
 */
class BinderParamBuilder extends ParamBuilder {

    /**
     * Initialize the builder
     */
    protected BinderParamBuilder(Messager messager, Element element) {
        super(messager, element);
    }


    @Override
    public void writeParamsToProxy(VariableElement param, ParamType paramType, MethodSpec.Builder methodBuilder) {
        if (param.asType().getKind() == TypeKind.ARRAY) {
            TypeMirror type = ((ArrayType) param.asType()).getComponentType();
            methodBuilder.beginControlFlow("if(" + param.getSimpleName() + " == null)")
                    .addStatement("data.writeInt(-1)")
                    .nextControlFlow("else")
                    .addStatement("data.writeInt(" + param.getSimpleName() + ".length)")
                    .beginControlFlow("for($T item : " + param.getSimpleName() + ")", type)
                    .addStatement("data.writeStrongBinder(item != null ? new $T(item)  : null)",
                            getStubClassName(type))
                    .endControlFlow()
                    .endControlFlow();
        } else {

            String binderName = param.getSimpleName() +"_binder";
            methodBuilder.addStatement("IBinder "+ binderName +" = null");
            methodBuilder.beginControlFlow("if (" + param.getSimpleName()  + " != null)");
            methodBuilder.beginControlFlow("synchronized (stubMap)");
            methodBuilder.addStatement(binderName + " = stubMap.get(" + param.getSimpleName()  +")");
            methodBuilder.beginControlFlow("if (" + binderName + " == null)");
            methodBuilder.addStatement(binderName + " = new $T("+ param.getSimpleName() +")", getStubClassName(param.asType()));
            methodBuilder.addStatement("stubMap.put(" +  param.getSimpleName() + ", " + binderName +")");
            methodBuilder.endControlFlow();
            methodBuilder.endControlFlow();
            methodBuilder.endControlFlow();

            methodBuilder.addStatement("data.writeStrongBinder(" + binderName + ")");
        }
    }

    @Override
    public void readResultsFromStub(TypeMirror resultType, MethodSpec.Builder methodBuilder) {
        if (resultType.getKind() == TypeKind.ARRAY) {
            TypeMirror type = ((ArrayType) resultType).getComponentType();
            methodBuilder.beginControlFlow("if(result == null)")
                    .addStatement("data.writeInt(-1)")
                    .nextControlFlow("else")
                    .addStatement("data.writeInt(result.length)")
                    .beginControlFlow("for($T item : result)", type)
                    .addStatement("data.writeStrongBinder(item != null ? new $T(item)  : null)",
                            getStubClassName(type))
                    .endControlFlow()
                    .endControlFlow();
        } else {
            methodBuilder.addStatement("reply.writeStrongBinder(result != null ? new $T(result)  : null)",
                    getStubClassName(resultType));

        }
    }


    @Override
    public void readResultsFromProxy(TypeMirror resultType, MethodSpec.Builder methodBuilder) {
        if (resultType.getKind() == TypeKind.ARRAY) {
            TypeMirror type = ((ArrayType) resultType).getComponentType();
            methodBuilder.addStatement("int length = reply.readInt()")
                    .beginControlFlow("if(length == -1)")
                    .addStatement("result = null")
                    .nextControlFlow("else")
                    .addStatement("result = new $T[length]", type)
                    .beginControlFlow("for(int i = 0; i < length; i++)")
                    .addStatement("result[i] = new $T(reply.readStrongBinder())", getProxyClassName(type))
                    .endControlFlow()
                    .endControlFlow();
        } else {
            String binderName = "result_binder";
            methodBuilder.addStatement("IBinder "+ binderName +" = reply.readStrongBinder()");
            methodBuilder.addStatement("result = null");
            methodBuilder.beginControlFlow("if("+ binderName +" != null)");
            methodBuilder.addStatement("result = new $T("+ binderName +")", getProxyClassName(resultType));
            methodBuilder.addStatement("(($T)result).setRemoterGlobalProperties(__global_properties)", getProxyClassName(resultType));
            methodBuilder.endControlFlow();
        }
    }

    @Override
    public void readOutResultsFromStub(VariableElement param, ParamType paramType, String paramName, MethodSpec.Builder methodBuilder) {
    }

    @Override
    public void writeParamsToStub(VariableElement param, ParamType paramType, String paramName, MethodSpec.Builder methodBuilder) {
        super.writeParamsToStub(param, paramType, paramName, methodBuilder);
        if (param.asType().getKind() == TypeKind.ARRAY) {
            TypeMirror type = ((ArrayType) param.asType()).getComponentType();
            methodBuilder.addStatement("int length = data.readInt()")
                    .beginControlFlow("if(length == -1)")
                    .addStatement(paramName + " = null")
                    .nextControlFlow("else")
                    .addStatement(paramName + " = new $T[length]", type)
                    .beginControlFlow("for(int i = 0; i < length; i++)")
                    .addStatement(paramName + "[i] = new $T(data.readStrongBinder())", getProxyClassName(type))
                    .endControlFlow()
                    .endControlFlow();
        } else {
            String binderName = paramName +"_binder";
            methodBuilder.addStatement("IBinder "+ binderName +" = data.readStrongBinder()");
            methodBuilder.addStatement(paramName + " = null");
            methodBuilder.beginControlFlow("if("+ binderName +" != null)");
            methodBuilder.addStatement(paramName + " = new $T("+ binderName +")", getProxyClassName(param.asType()));
            methodBuilder.endControlFlow();
        }
    }

    @Override
    public void readOutParamsFromProxy(VariableElement param, ParamType paramType, MethodSpec.Builder methodBuilder) {
    }

    /**
     * Returns the {@link ClassName} for the Stub
     */
    private ClassName getStubClassName(TypeMirror param) {
        Element element = getBindingManager().getElement(param.toString());
        return ClassName.get(getPackage(element).getQualifiedName().toString(), element.getSimpleName() + ClassBuilder.STUB_SUFFIX);
    }

    /**
     * Returns the {@link ClassName} for the Proxy
     */
    private ClassName getProxyClassName(TypeMirror param) {
        Element element = getBindingManager().getElement(param.toString());
        return ClassName.get(getPackage(element).getQualifiedName().toString(), element.getSimpleName() + ClassBuilder.PROXY_SUFFIX);
    }

}

File: remoter/src/main/java/remoter/compiler/kbuilder/BinderParamBuilder.kt
package remoter.compiler.kbuilder

import com.google.auto.common.MoreElements
import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.FunSpec
import javax.lang.model.element.Element
import javax.lang.model.element.ExecutableElement
import javax.lang.model.element.VariableElement
import javax.lang.model.type.ArrayType
import javax.lang.model.type.TypeKind
import javax.lang.model.type.TypeMirror


/**
 * A [ParamBuilder] for Binder type parameters
 */
internal class BinderParamBuilder(remoterInterfaceElement: Element, bindingManager: KBindingManager) : ParamBuilder(remoterInterfaceElement, bindingManager) {


    override fun writeParamsToProxy(param: VariableElement, paramType: ParamType, methodBuilder: FunSpec.Builder) {
        if (param.asType().kind == TypeKind.ARRAY) {
            val type = (param.asType() as ArrayType).componentType
            if (param.isNullable()) {
                methodBuilder.beginControlFlow("if(" + param.simpleName + " == null)")
                        .addStatement("$DATA.writeInt(-1)")
                        .nextControlFlow("else")
                        .addStatement("$DATA.writeInt(" + param.simpleName + ".size)")
                        .beginControlFlow("for(__item in " + param.simpleName + ")", type)
                        .addStatement("$DATA.writeStrongBinder(if(__item != null)  %T(__item)  else null)",
                                getStubClassName(type))
                        .endControlFlow()
                        .endControlFlow()
            } else {
                methodBuilder.addStatement("$DATA.writeInt(" + param.simpleName + ".size)")
                        .beginControlFlow("for(__item in " + param.simpleName + ")", type)
                        .addStatement("$DATA.writeStrongBinder(if(__item != null)  %T(__item)  else null)",
                                getStubClassName(type))
                        .endControlFlow()
            }
        } else {
            val binderName = param.simpleName.toString() + "_binder"
            methodBuilder.addStatement("var $binderName:IBinder?")
            if (param.isNullable()) {
                methodBuilder.beginControlFlow("if (" + param.simpleName + " != null)")
            }
            methodBuilder.beginControlFlow("synchronized (stubMap)")
            methodBuilder.addStatement(binderName + " = stubMap.get(" + param.simpleName + ")")
            methodBuilder.beginControlFlow("if ($binderName == null)")
            methodBuilder.addStatement(binderName + " = %T(" + param.simpleName + ")", getStubClassName(param.asType()))
            methodBuilder.addStatement("stubMap.put(" + param.simpleName + ", " + binderName + ")")
            methodBuilder.endControlFlow()
            methodBuilder.endControlFlow()
            if (param.isNullable()) {
                methodBuilder.endControlFlow()
                methodBuilder.beginControlFlow("else")
                methodBuilder.addStatement("$binderName = null")
                methodBuilder.endControlFlow()
            }
            methodBuilder.addStatement("$DATA.writeStrongBinder($binderName)")
        }
    }

    override fun readResultsFromStub(methodElement: ExecutableElement, resultType: TypeMirror, methodBuilder: FunSpec.Builder) {
        if (resultType.kind == TypeKind.ARRAY) {
            val type = (resultType as ArrayType).componentType
            if (methodElement.isNullable()) {
                methodBuilder.beginControlFlow("if($RESULT == null)")
                        .addStatement("$DATA.writeInt(-1)")
                        .nextControlFlow("else")
                        .addStatement("$DATA.writeInt($RESULT.size)")
                        .beginControlFlow("for(item in $RESULT)")
                        .addStatement("$DATA.writeStrongBinder( if(item != null) %T(item)  else null)",
                                getStubClassName(type))
                        .endControlFlow()
                        .endControlFlow()
            } else {
                methodBuilder.addStatement("$DATA.writeInt($RESULT.size)")
                        .beginControlFlow("for(item in $RESULT)")
                        .addStatement("$DATA.writeStrongBinder( if(item != null) %T(item)  else null)",
                                getStubClassName(type))
                        .endControlFlow()
            }
        } else {
            if (methodElement.isNullable()) {
                methodBuilder.addStatement("$REPLY.writeStrongBinder(if($RESULT != null)  %T($RESULT)  else null)",
                        getStubClassName(resultType))
            } else {
                methodBuilder.addStatement("$REPLY.writeStrongBinder(%T($RESULT))",
                        getStubClassName(resultType))

            }
        }
    }

    override fun readResultsFromProxy(methodType: ExecutableElement, methodBuilder: FunSpec.Builder) {
        val resultType = methodType.getReturnAsKotlinType()
        val resultMirror = methodType.getReturnAsTypeMirror()
        if (resultMirror.kind == TypeKind.ARRAY) {
            val type = (resultMirror as ArrayType).componentType
            methodBuilder.addStatement("val ___result_length = $REPLY.readInt()")
                    .beginControlFlow("if(___result_length == -1)")
            if (methodType.isNullable()) {
                methodBuilder.addStatement("$RESULT = null")
            } else {
                methodBuilder.addStatement("throw %T(\"Unexpected null result\")", NullPointerException::class)
            }

            methodBuilder.nextControlFlow("else")
                    .beginControlFlow("$RESULT = Array(___result_length)")
                    .addStatement("%T($REPLY.readStrongBinder())", getProxyClassName(type))
                    .endControlFlow()
                    .endControlFlow()
        } else {
            val binderName = "__result_binder"
            methodBuilder.addStatement("val $binderName = $REPLY.readStrongBinder()")
            if (resultType.isNullable) {
                methodBuilder.addStatement("$RESULT = null")
            }
            methodBuilder.beginControlFlow("if($binderName != null)")
            methodBuilder.addStatement("$RESULT = %T($binderName)", getProxyClassName(resultMirror))
            methodBuilder.addStatement("$RESULT.setRemoterGlobalProperties(__global_properties)")
            methodBuilder.endControlFlow()
            methodBuilder.beginControlFlow("else")
            if (!resultType.isNullable) {
                methodBuilder.addStatement("throw %T(\"Unexpected null result\")", NullPointerException::class.java)
            }
            methodBuilder.endControlFlow()

        }
    }

    override fun readOutResultsFromStub(param: VariableElement, paramType: ParamType, paramName: String, methodBuilder: FunSpec.Builder) {}
    override fun writeParamsToStub(methodType: ExecutableElement, param: VariableElement, paramType: ParamType, paramName: String, methodBuilder: FunSpec.Builder) {
        super.writeParamsToStub(methodType, param, paramType, paramName, methodBuilder)
        if (param.asType().kind == TypeKind.ARRAY) {
            val type = param.asKotlinType()
            methodBuilder.addStatement("val length = $DATA.readInt()")
                    .beginControlFlow("if(length == -1)")
            if (param.isNullable()) {
                methodBuilder.addStatement("$paramName = null")
            } else {
                methodBuilder.addStatement("throw NullPointerException(\"Not expecting null\")")
            }
            methodBuilder.nextControlFlow("else")
                    .beginControlFlow("$paramName = %T(length) ", type.copy(false))
                    .addStatement("%T($DATA.readStrongBinder())", getProxyClassName((param.asType() as ArrayType).componentType))
                    .endControlFlow()
                    .endControlFlow()
        } else {
            val binderName = paramName + "_binder"
            methodBuilder.addStatement("val $binderName = $DATA.readStrongBinder()")
            if (param.isNullable()) {
                methodBuilder.addStatement("$paramName = null")
            }
            methodBuilder.beginControlFlow("if($binderName != null)")
            methodBuilder.addStatement("$paramName =  %T($binderName)", getProxyClassName(param.asType()))
            methodBuilder.endControlFlow()
            methodBuilder.beginControlFlow("else")
            if (!param.isNullable()) {
                methodBuilder.addStatement("throw %T(\"Not expecting null\")", NullPointerException::class)
            }
            methodBuilder.endControlFlow()

        }
    }

    override fun readOutParamsFromProxy(param: VariableElement, paramType: ParamType, methodBuilder: FunSpec.Builder) {}
    /**
     * Returns the [ClassName] for the Stub
     */
    private fun getStubClassName(param: TypeMirror): ClassName {
        return getStubClassName(param.toString())
    }

    /**
     * Returns the [ClassName] for the Proxy
     */
    private fun getProxyClassName(param: TypeMirror): ClassName {
        return getProxyClassName(param.toString())
    }

    /**
     * Returns the [ClassName] for the Stub
     */
    private fun getStubClassName(name: String): ClassName {
        val element = bindingManager.getElement(name)
        return ClassName(MoreElements.getPackage(element).qualifiedName.toString(), element.simpleName.toString() + KClassBuilder.STUB_SUFFIX)
    }

    /**
     * Returns the [ClassName] for the Proxy
     */
    private fun getProxyClassName(name: String): ClassName {
        val element = bindingManager.getElement(name)
        return ClassName(MoreElements.getPackage(element).qualifiedName.toString(), element.simpleName.toString() + KClassBuilder.PROXY_SUFFIX)
    }
}